### ISSUE TITLE
Multi-GPU training

### ISSUE THREAD

**Issue by Danyache** [PyTorchLightning/pytorch-lightning#9106]
### Discussed in https://github.com/PyTorchLightning/pytorch-lightning/discussions/9092

<div type='discussions-op-text'>

<sup>Originally posted by **Danyache** August 24, 2021</sup>
Hi!
Thanks for your lib, it's quite impressive. I'm trying to train a big model, using pytorch lightning and have a questiong. I want to train the model with some big batchsize, which is too big to fit on one GPU, but also I want to calculate cross-entropy loss over all the batch. As I've read from the docs, there is a possibility to do so with "training_step_end". However it's only used with dp training. Is it any possibility to calculate cross entropy loss over all the batches using ddp training? And if it's not -- is there a possibility to use apex with dp training? Because when I use torch backend for fp16 training and using FuseAdam from apex -- there s no layers with fp16 at all. So there are 4 main questions:
1. Is it possible to calculate cross entropy loss over all the batches with ddp training?
2. If it's not -- is it possible to use apex and dp training?
3. And if that's not possible too -- how to make my fp16 training to be really the fp16 training or maybe there are some other ideas of how to speedup training and reduce the memory usage
4. Is it possible to use activation checkpointing with dp training?</div>

**Comment by anhnht3** [comment]
https://pytorch-lightning.readthedocs.io/en/latest/common/trainer.html#accumulate-grad-batches

**Comment by tchaton** [comment]
Dear @Danyache,

I have recently seen an implementation for such use-case but can't find it back.

An alternative would be for you to implement it. Lightning provides an all_gather methods which supports backward propagation.

You could use

```py
F.cross_entropy(self.all_gather(logits), self.all_gather(labels))
 ```

Note: This would compute the loss world_size times and your learning rate would have to account for this.

However, you might find a better implementation out there.

Best,
T.C
